src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+                    G_GNUC_UNUSED const char *name,
+                    G_GNUC_UNUSED const char *arch,
+                    void *cbdata,
+                    GError **err)
+{
+    assert(pkg && *pkg == NULL);
+    assert(!err || *err == NULL);
+
+    cr_CbData *cb_data = cbdata;
+
+    gpointer pkey = NULL;
+    gpointer pval = NULL;
+    if (g_hash_table_lookup_extended(cb_data->in_progress_pkgs, pkgId, &pkey, &pval)) {
+        // package was already encountred in some other metadata type
+
+        if (pval == NULL) {
+            assert(0);
+            //this pkg was already fully handled
+            //TODO(amatej): maybe error?
+        } else {
+            *pkg = (cr_Package *) pval;
+        }
+    } else {
+        // we are handling never before seen package
+
+        if (cb_data->newpkgcb) {
+            // call user specified new function if provided
+            if (cb_data->newpkgcb(pkg, pkgId, name, arch, cb_data->newpkgcb_data, err)) {
+                return CR_CB_RET_ERR;
+            }
+        } else {
+            *pkg = cr_package_new();
+        }
+        g_hash_table_insert(cb_data->in_progress_pkgs, g_strdup(pkgId), *pkg);
+    }
+
+    return CR_CB_RET_OK;
+}
+
+static int
+filelists_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->filelists_count++;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_FIL;
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+other_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->other_count++;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_OTH;
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+primary_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->primary_count++;
+
+    gpointer pkey = NULL;
+    gpointer pval = NULL;
+    if (g_hash_table_lookup_extended(cb_data->in_progress_pkgs, pkg->pkgId, &pkey, &pval)) {
+        // package was already encountred in some other metadata type
+
+        if (pval == NULL) {
+            //this pkg was already fully handled
+            //TODO(amatej): maybe error?
+            assert(0);
+        } else {
+            cr_package_copy_into(pkg, pval);
+            cr_package_free(pkg);
+            pkg = pval;
+        }
+    } else {
+        // we are handling never before seen package
+
+        if (cb_data->newpkgcb) {
+            // call user specified new function if provided and copy q
+            cr_Package *user_created_pkg;
+            if (cb_data->newpkgcb(&user_created_pkg, pkg->pkgId, NULL, NULL, cb_data->newpkgcb_data, err)) {
+                return CR_CB_RET_ERR;
+            } else {
+                cr_package_copy_into(pkg, user_created_pkg);
+                cr_package_free(pkg);
+                pkg = user_created_pkg;
+            }
```

Comment:
There's probably a good reason we need to copy the metadata instead of parsing it into the existing package - probably something to do with how primary.xml package records have a different structure from the others which don't include the pkgid as an attribute, yes?  Or something entirely unrelated?